src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -0,0 +1,1254 @@
+      group_snap_id);
+  m_local_group_snaps.push_back(local_snap);
+
+  auto comp = create_rados_callback(
+      new LambdaContext([this, group_snap_id, on_finish](int r) {
+        handle_create_mirror_snapshot(r, group_snap_id, on_finish);
+        }));
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::group_snap_set(&op, local_snap);
+  r = m_local_io_ctx.aio_operate(
+      librbd::util::group_header_name(m_local_group_id), comp, &op);
+  ceph_assert(r == 0);
+  locker.unlock();
+  comp->release();
+
+  // if m_replayer in the ImageReplayer is null this cannot be forwarded.
+  // May be we should retry this setting in the validate_image_snaps_sync_complete().
+  // Same for image_replayer->prune_snapshot(); setting actually!!!!
+  set_image_replayer_limits("", snap);
+}
+
+template <typename I>
+void Replayer<I>::handle_create_mirror_snapshot(
+    int r, const std::string &remote_group_snap_id, Context *on_finish) {
+  dout(10) << remote_group_snap_id << ", r=" << r << dendl;
+
+  on_finish->complete(0);
+}
+
+template <typename I>
+void Replayer<I>::mirror_snapshot_complete(
+    const std::string &remote_group_snap_id,
+    cls::rbd::ImageSnapshotSpec *spec,
+    Context *on_finish) {
+
+  ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
+
+  auto itr = std::find_if(
+      m_remote_group_snaps.begin(), m_remote_group_snaps.end(),
+      [remote_group_snap_id](const cls::rbd::GroupSnapshot &s) {
+      return s.id == remote_group_snap_id;
+      });
+
+  ceph_assert(itr != m_remote_group_snaps.end());
+  auto itl = std::find_if(
+      m_local_group_snaps.begin(), m_local_group_snaps.end(),
+      [remote_group_snap_id](const cls::rbd::GroupSnapshot &s) {
+      return s.id == remote_group_snap_id;
+      });
+  if (itr->snaps.size() != 0) {
+    // update the group snap with snap spec
+    itl->snaps.push_back(*spec);
+  }
+
+  if (itr->snaps.size() == itl->snaps.size()) {
+    itl->state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+
+    auto comp = create_rados_callback(
+        new LambdaContext([this, remote_group_snap_id, on_finish](int r) {
+          handle_mirror_snapshot_complete(r, remote_group_snap_id, on_finish);
+          }));
+
+    librados::ObjectWriteOperation op;
+    librbd::cls_client::group_snap_set(&op, *itl);
+    int r = m_local_io_ctx.aio_operate(
+        librbd::util::group_header_name(m_local_group_id), comp, &op);
+    ceph_assert(r == 0);
+    comp->release();
+  } else {
+    on_finish->complete(0);
+  }
+
+  dout(10) << "local group snap info: "
+           << "id: " << itl->id
+           << ", name: " << itl->name
+           << ", state: " << itl->state
+           << ", snaps.size: " << itl->snaps.size()
+           << dendl;
+
+  return;
+}
+
+template <typename I>
+void Replayer<I>::handle_mirror_snapshot_complete(
+    int r, const std::string &remote_group_snap_id, Context *on_finish) {
+  dout(10) << remote_group_snap_id << ", r=" << r << dendl;
+
+  on_finish->complete(0);
```

Comment:
This doesn't seem right. Should this be,
```
on_finish->complete(r)
```
Currently,  any error from `group_snap_set` is not being handled.